crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2391,75 +2434,100 @@ impl<'db> TypeInferenceBuilder<'db> {
-                    Some(self.lookup_name(name))
-                } else {
-                    None
-                };
-                let ty = bindings_ty(self.db, definitions, unbound_ty);
+        let ty = bindings_ty(self.db, definitions, unbound_ty);
+        if ty.is_unbound() {
+            self.add_diagnostic(
+                name.into(),
+                "unresolved-reference",
+                format_args!("Name `{id}` used when not defined"),
+            );
+        } else if ty.may_be_unbound(self.db) {
+            self.add_diagnostic(
+                name.into(),
+                "possibly-unresolved-reference",
+                format_args!("Name `{id}` used when possibly not defined"),
+            );
+        }
 
-                if ty.is_unbound() {
-                    self.add_diagnostic(
-                        name.into(),
-                        "unresolved-reference",
-                        format_args!("Name `{id}` used when not defined"),
-                    );
-                } else if ty.may_be_unbound(self.db) {
-                    self.add_diagnostic(
-                        name.into(),
-                        "possibly-unresolved-reference",
-                        format_args!("Name `{id}` used when possibly not defined"),
-                    );
-                }
+        ty
+    }
 
-                ty
-            }
+    fn infer_name_expression(&mut self, name: &ast::ExprName) -> Type<'db> {
+        match name.ctx {
+            ExprContext::Load => self.infer_name_load(name),
             ExprContext::Store | ExprContext::Del => Type::None,
             ExprContext::Invalid => Type::Unknown,
         }
     }
 
-    fn infer_attribute_expression(&mut self, attribute: &ast::ExprAttribute) -> Type<'db> {
+    /// Infer the type of a [`ast::ExprAttribute`] expression, assuming a load context.
+    fn infer_attribute_load(&mut self, attribute: &ast::ExprAttribute) -> Type<'db> {
         let ast::ExprAttribute {
             value,
             attr,
             range: _,
-            ctx,
+            ctx: _,
         } = attribute;
 
         let value_ty = self.infer_expression(value);
-        let member_ty = value_ty.member(self.db, &ast::name::Name::new(&attr.id));
+        let member_ty = value_ty.member(self.db, &Name::new(&attr.id));
+
+        member_ty
+    }
+
+    fn infer_attribute_expression(&mut self, attribute: &ast::ExprAttribute) -> Type<'db> {
+        let ast::ExprAttribute {
+            value,
+            attr: _,
+            range: _,
+            ctx,
+        } = attribute;
 
         match ctx {
-            ExprContext::Load => member_ty,
-            ExprContext::Store | ExprContext::Del => Type::None,
-            ExprContext::Invalid => Type::Unknown,
+            ExprContext::Load => self.infer_attribute_load(attribute),
+            ExprContext::Store | ExprContext::Del => {
+                self.infer_expression(value);
```

Comment:
I assume these still need to happen, even though we're not using the result, since they mutate the inference builder.

Reply:
Yeah, we currently maintain an invariant that every expression gets a type inferred for it, so we have to infer types even if we won't use them. (The theory is that in future an LSP may want to query a type on hover for any expression.)